src/threading/thread.cpp

```diff
@@ -0,0 +1,350 @@
+#else
+# ifdef _WIN32
+	if (!running) return;
+	WaitForSingleObject(thread, INFINITE);
+# else // pthread
+	void *status;
+	if (!started) return;
+	int ret = pthread_join(thread, &status);
+	assert(!ret);
+	UNUSED(ret);
+	started = false;
+# endif
+#endif
+}
+
+
+bool Thread::start()
+{
+	if (running) return false;
+	request_stop = false;
+
+#if __cplusplus >= 201103L
+	MutexAutoLock l(continue_mutex);
+	thread = new std::thread(theThread, this);
+#else
+# ifdef _WIN32
+	MutexAutoLock l(continue_mutex);
+#  ifdef _WIN32_WCE
+	thread = CreateThread(NULL, 0, theThread, this, 0, &thread_id);
+#  else
+	thread = (HANDLE) _beginthreadex(NULL, 0, theThread, this, 0, &thread_id);
+#  endif
+	if (!thread) return false;
+# else
+	int status;
+
+	pthread_attr_t attr;
+	pthread_attr_init(&attr);
+
+	MutexAutoLock l(continue_mutex);
+
+	status = pthread_create(&thread, &attr, theThread, this);
+	pthread_attr_destroy(&attr);
+
+	if (status) return false;
+# endif
+#endif
+
+#if __cplusplus < 201103L
+	// Wait until running
+	while (!running) {
+# ifdef _WIN32
+		Sleep(1);
+	}
+# else
+		struct timespec req, rem;
+		req.tv_sec = 0;
+		req.tv_nsec = 1000000;
+		nanosleep(&req, &rem);
+	}
+	started = true;
+# endif
+#endif
+	return true;
+}
+
+
+bool Thread::kill()
+{
+#ifdef _WIN32
+	if (!running) return false;
+	TerminateThread(THREAD_HANDLE, 0);
+	CloseHandle(THREAD_HANDLE);
+#else
+	if (!running) {
+		wait();
+		return false;
+	}
+# ifdef __ANDROID__
+	pthread_kill(THREAD_HANDLE, SIGKILL);
+# else
+	pthread_cancel(THREAD_HANDLE);
+# endif
+	wait();
+#endif
+#if __cplusplus >= 201103L
+	delete thread;
+#endif
+	running = false;
+	return true;
+}
+
+
+bool Thread::isSameThread()
+{
+#if __cplusplus >= 201103L
+	return thread->get_id() == std::this_thread::get_id();
+#else
+# ifdef _WIN32
+	return GetCurrentThreadId() == thread_id;
+# else
+	return pthread_equal(pthread_self(), thread);
+# endif
+#endif
+}
+
+
+#if __cplusplus >= 201103L
+void Thread::theThread(Thread *th) {
+#else
+# if defined(_WIN32_WCE)
+DWORD WINAPI
+# elif defined(_WIN32)
+UINT __stdcall
+# else
+void*
+# endif
+Thread::theThread(void *param) {
+	Thread *th = static_cast<Thread *>(param);
+#endif
+	th->running = true;
+
+	th->setName();
+	th->retval = th->run();
+
+	th->running = false;
+#if __cplusplus < 201103L
+# ifdef _WIN32
+	CloseHandle(th->thread);
+# endif
+	return NULL;
+#endif
+}
+
+
+void Thread::setName(const std::string &name)
+{
+#if defined(linux) || defined(__linux)
+	/* It would be cleaner to do this with pthread_setname_np,
+	 * which was added to glibc in version 2.12, but some major
+	 * distributions are still runing 2.11 and previous versions.
+	 */
+	prctl(PR_SET_NAME, name.c_str());
+#elif defined(__FreeBSD__) || defined(__OpenBSD__)
+	pthread_set_name_np(pthread_self(), name.c_str());
+#elif defined(__NetBSD__)
+	pthread_setname_np(pthread_self(), name.c_str());
+#elif defined(__APPLE__)
+	pthread_setname_np(name.c_str());
+#elif defined(_MSC_VER)
```

Comment:
Yep, Windows itself doesn't actually support thread names, but the MSVC debugger does (this is why the next condition catches Windows).  I'll add that to the comment.